### PR TITLE
WebApi file routes

### PR DESCRIPTION
**Pull request by gumbarros**

### CODE REVIEW

File: src/JJMasterData.Core/WebComponents/JJDataPanel.cs [hunk 1]
```diff
@@ -241,9 +241,9 @@ public Hashtable GetFormValues()
             if (!string.IsNullOrEmpty(criptPkval))
             {
                 string parsedPkval = Cript.Descript64(criptPkval);
-                var filters = DataHelper.GetPkValues(FormElement, parsedPkval, '|');
+                var filters = new Hashtable(DataHelper.GetPkValues(FormElement, parsedPkval, '|'));
```
Comment (LucioPelinson) [inline src/JJMasterData.Core/WebComponents/JJDataPanel.cs]: oops

File: src/JJMasterData.WebApi/Controllers/MasterApiController.cs [hunk 2]
```diff
@@ -11,20 +11,20 @@ namespace JJMasterData.WebApi.Controllers;
 
 [Authorize]
 [ApiController]
+[Route("masterApi/{elementName}")]
```
Comment (LucioPelinson) [inline src/JJMasterData.WebApi/Controllers/MasterApiController.cs:14]: ok

File: src/JJMasterData.Core/WebComponents/JJDataPanel.cs [hunk 3]
```diff
@@ -241,9 +241,9 @@ public Hashtable GetFormValues()
...
                 var entityRepository = FieldManager.Expression.EntityRepository;
-                tempvalues = entityRepository.GetFields(FormElement, filters);
+                tempvalues = new Hashtable(entityRepository.GetFields(FormElement, filters));
```
Comment (LucioPelinson) [inline src/JJMasterData.Core/WebComponents/JJDataPanel.cs]: GetFields already return a Hashtable, revert this block, please!

File: src/JJMasterData.WebApi/Controllers/MasterApiController.cs [hunk 4]
```diff
@@ -36,86 +36,106 @@ public MasterApiController(MasterApiService service)
...
+        var fileStream = _service.GetDictionaryFile(elementName, id, fieldName, fileName);
+
+        return File(fileStream, "application/octet-stream", fileName);
+    }
 
     [HttpPost]
-    [Route("masterApi/{elementName}/action/{actionName?}/{fieldName?}")]
-    public HttpResponseMessage Action(string elementName, [FromBody]Hashtable paramValues, string actionName, string fieldName = "")
+    [Route("{id}/file/{fieldName}")]
+    public IActionResult PostFile(string elementName, string fieldName, string id, IFormFile file)
```
Comment (LucioPelinson) [inline src/JJMasterData.WebApi/Controllers/MasterApiController.cs]: Swagger documentation is missing

File: src/JJMasterData.WebApi/Services/MasterApiService.cs [hunk 5]
```diff
@@ -459,45 +439,135 @@ private FormService GetFormService(Metadata metadata)
         
         return service;
     }
+    public FileStream GetDictionaryFile(string elementName, string id, string fieldName, string fileName)
+    {
+        var formElement = GetFormElement(elementName);
+
+        var folderPath = formElement.Fields.First(f => f.Name == fieldName).DataFile.FolderPath;
+
+        var path = Path.Combine(folderPath, id);
```
Comment (LucioPelinson) [inline src/JJMasterData.WebApi/Services/MasterApiService.cs]: Please, use FormFilePathBuilder class to build a file path

File: src/JJMasterData.WebApi/Services/MasterApiService.cs [hunk 6]
```diff
@@ -459,45 +439,135 @@ private FormService GetFormService(Metadata metadata)
...
+    /// Compares the values of the fields received with those sent to the bank, returning different records
     /// </summary>
     /// <remarks>
-    /// Isso acontece devido as triggers ou os valores 
-    /// retornados nos metodos de set (id autoNum) por exemplo
+    /// This happens due to triggers or values
+    /// returned in set methods (id autoNum) for example
     /// </remarks>
-    private Hashtable? GetDiff(Hashtable original, Hashtable result, ApiSettings api)
+    private static Hashtable? GetDiff(Hashtable original, Hashtable result, ApiSettings api)
```
Comment (LucioPelinson) [inline src/JJMasterData.WebApi/Services/MasterApiService.cs]: Remove static... "I hate this on Rider"

File: src/JJMasterData.Commons/Dao/Entity/Factory.cs [hunk 7]
```diff
@@ -106,8 +97,8 @@ internal BaseProvider Provider
     ///<inheritdoc cref="IEntityRepository.GetElementFromTable(string)"/>
     public Element GetElementFromTable(string tableName) => Provider.GetElementFromTable(tableName);
 
-    ///<inheritdoc cref="IEntityRepository.GetListFieldsAsText(Element, Hashtable, string, int, int, bool, string)"/>
-    public string GetListFieldsAsText(Element element, Hashtable filters, string orderBy, int recordsPerPage, int currentPage, bool showLogInfo, string delimiter = "|") =>
-        Provider.GetListFieldsAsText(element, filters, orderBy, recordsPerPage, currentPage, showLogInfo, delimiter);
+    ///<inheritdoc cref="IEntityRepository.GetListFieldsAsText(Element,IDictionary,string,int,int,bool,string)"/>
+    public string GetListFieldsAsText(Element element, IDictionary filters, string orderBy, int recordsPerPage, int currentPage, bool showLogInfo, string delimiter = "|") =>
+        Provider.GetListFieldsAsText(element, new Hashtable(filters), orderBy, recordsPerPage, currentPage, showLogInfo, delimiter);
```
Comment (LucioPelinson) [inline src/JJMasterData.Commons/Dao/Entity/Factory.cs:102]: Why didn't it end?

File: src/JJMasterData.Commons/Dao/Entity/Factory.cs [hunk 8]
```diff
@@ -43,32 +34,32 @@ internal BaseProvider Provider
         }
     }
 
-    ///<inheritdoc cref="IEntityRepository.Insert(Element, Hashtable)"/>
-    public void Insert(Element element, Hashtable values) => Provider.Insert(element, values);
+    ///<inheritdoc cref="IEntityRepository.Insert(Element, IDictionary)"/>
+    public void Insert(Element element, IDictionary values) => Provider.Insert(element, new Hashtable(values));
```
Comment (LucioPelinson) [inline src/JJMasterData.Commons/Dao/Entity/Factory.cs:38]: I understand your hatred of Hashtable, but take it easy!

File: src/JJMasterData.WebApi/Services/MasterApiService.cs [hunk 9]
```diff
@@ -99,73 +97,60 @@ public MasterApiListResponse GetListFields(string elementName, int pag, int regp
 
         return listRet;
     }
-
-    public List<ResponseLetter> SetFields(Hashtable[] listParam, string elementName, bool replace = false)
+    public IEnumerable<ResponseLetter> SetFields(IEnumerable<Hashtable> paramsList, string elementName, bool replace = false)
     {
-        if (listParam == null)
-            throw new ArgumentNullException(nameof(listParam));
+        if (paramsList == null)
```
Comment (LucioPelinson) [inline src/JJMasterData.WebApi/Services/MasterApiService.cs:104]: good

File: src/JJMasterData.WebApi/Controllers/MasterApiController.cs [hunk 10]
```diff
@@ -36,86 +36,106 @@ public MasterApiController(MasterApiService service)
...
-
-
+    
     [HttpPost]
-    [Route("masterApi/{elementName}")]
-    public ActionResult<ResponseLetter> Post([FromBody]Hashtable[] listParam, string elementName, bool replace = false)
+    public ActionResult<ResponseLetter> Post([FromBody] Hashtable[] listParam, string elementName, bool replace = false)
     {
-        return GetResponseMessage(_service.SetFields(listParam, elementName, replace));
+        return GetResponseMessage(new List<ResponseLetter>(_service.SetFields(listParam, elementName, replace)));
```
Comment (LucioPelinson) [inline src/JJMasterData.WebApi/Controllers/MasterApiController.cs]: ?

File: src/JJMasterData.WebApi/Services/MasterApiService.cs [hunk 11]
```diff
@@ -99,73 +97,60 @@ public MasterApiListResponse GetListFields(string elementName, int pag, int regp
...
             throw new UnauthorizedAccessException();
 
         var formService = GetFormService(dictionary);
-        var listRet = new List<ResponseLetter>();
-        foreach (Hashtable values in listParam)
+
+        foreach (var values in paramsList)
         {
-            ResponseLetter ret = replace ?
+            yield return replace ?
```
Comment (LucioPelinson) [inline src/JJMasterData.WebApi/Services/MasterApiService.cs:115]: I prefer the other way

Comment (LucioPelinson) [reply src/JJMasterData.WebApi/Services/MasterApiService.cs:115]: Relax with Rider...

**Review by LucioPelinson — CHANGES_REQUESTED** [review]

File: src/JJMasterData.WebApi/Services/MasterApiService.cs (hunk 11 above)
Comment (gumbarros) [reply src/JJMasterData.WebApi/Services/MasterApiService.cs:115]: Increbily this is not a Rider suggestion, I was studying about the `yield` keyword, makes a lot of sense to me.
https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/statements/yield

This is what made me a big fan of `yield`
image

When using `yield`, we no longer need the boilerplate `ret` variable we use on foreach.

File: src/JJMasterData.WebApi/Controllers/MasterApiController.cs (hunk 10 above)
Comment (gumbarros) [reply src/JJMasterData.WebApi/Controllers/MasterApiController.cs]: My mistake

File: src/JJMasterData.WebApi/Services/MasterApiService.cs (hunk 6 above)
Comment (gumbarros) [reply src/JJMasterData.WebApi/Services/MasterApiService.cs]: Yeah, I accepted just for the sake of Rider stopping suggesting, maybe it's better to add to the team-shared DotFile to ignore this.

File: src/JJMasterData.WebApi/Services/MasterApiService.cs (hunk 5 above)
Comment (gumbarros) [reply src/JJMasterData.WebApi/Services/MasterApiService.cs]: Nice suggestion

File: src/JJMasterData.Core/WebComponents/JJDataPanel.cs (hunk 1 above)
Comment (gumbarros) [reply src/JJMasterData.Core/WebComponents/JJDataPanel.cs]: Not a error, this Hashtable constructor expects a IDictionary. I can explain better the intention tomorrow, but I understand the mistake when not padronizing everything

File: src/JJMasterData.WebApi/Services/MasterApiService.cs (hunk 11 above)
Comment (gumbarros) [reply src/JJMasterData.WebApi/Services/MasterApiService.cs:115]: Example of `foreach` without the boilerplate `ret` variable

```cs
var numbers = GetEvenNumbers(5);
foreach (int i in numbers)
{
    Console.WriteLine($"{i}");
}

IEnumerable<int> GetEvenNumbers(int upto)
{
    for (int i = 0; i <= upto; i += 2)
    {
        yield return i;
    }
}
```

In Portguese `yield return` means something like "Produza para o meu return"

**Comment by gumbarros** [comment]
@LucioPelinson we can talk better tomorrow, I made a mistake with the `IDictionary` thing. I agree with you **it's better to padronize everything**. The reason of the `var a = new Hashtable(IDictionary variable)` is because if I convert variable a to IDictionary, the entire library will collapse like a virus requiring `IDictionary`.

File: src/JJMasterData.WebApi/Services/MasterApiService.cs (hunk 11 above)
Comment (LucioPelinson) [reply src/JJMasterData.WebApi/Services/MasterApiService.cs:115]: This put a bug in my mind! But I understand now. tks for a example

File: src/JJMasterData.Commons/Dao/Entity/Factory.cs (hunk 8 above)
Comment (gumbarros) [reply src/JJMasterData.Commons/Dao/Entity/Factory.cs:38]: I will adress this in a new PR

**Comment by gumbarros** [comment]
@LucioPelinson  ready for review again.

**Comment by gumbarros** [comment]
Important observations:
1. I will address the IDictionary chaos in another PR
2. Fixed a bug where API don't startup, because of both Swagger and Logger

**Comment by gumbarros** [comment]
image
@LucioPelinson  this is "fino do fino"

**Pull request merged**